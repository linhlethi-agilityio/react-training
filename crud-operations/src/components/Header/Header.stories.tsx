import type { Meta, StoryObj } from '@storybook/react';

// Components
import Header from '.';

const meta: Meta<typeof Header> = {
  component: Header,
};

export default meta;
type Story = StoryObj<typeof Header>;

export const Default: Story = {
  args: {
    isClosedSideBar: false,
    onToggleSideBar: () => null,
    onSearch: () => null,
  },
};
