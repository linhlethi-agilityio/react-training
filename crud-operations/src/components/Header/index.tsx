import { memo, useState } from 'react';
import { Button, Flex, Icon } from '@chakra-ui/react';

// Icons
import { BellIcon, CaretIcon } from '@icons';

// Components
import { SearchInput } from '@components';

interface HeaderProps {
  isClosedSideBar?: boolean;
  onToggleSideBar?: () => void;
  onSearch: (keyword: string) => void;
}

const Header = ({ isClosedSideBar, onToggleSideBar, onSearch }: HeaderProps) => {
  const [searchValue, setSearchValue] = useState<string>('');

  const handleOnChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const textInputted = e.target.value;

    setSearchValue(textInputted);
    onSearch(textInputted);
  };

  return (
    <Flex justifyContent="space-between" alignItems="center" padding={3}>
      <Button data-testid="toggle-sidebar-button" variant="ghost" onClick={onToggleSideBar}>
        <Icon
          as={CaretIcon}
          boxSize={18}
          transform={isClosedSideBar ? 'rotate(180deg)' : 'rotate(0deg)'}
          transition="1s"
        />
      </Button>
      <Flex gap={27} alignItems="center">
        <SearchInput placeholder="Search..." onChange={handleOnChange} value={searchValue} />
        <Button variant="ghost">
          <BellIcon />
        </Button>
      </Flex>
    </Flex>
  );
};

export default memo(Header);
