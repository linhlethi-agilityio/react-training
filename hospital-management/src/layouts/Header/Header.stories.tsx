import { ComponentMeta, ComponentStory } from '@storybook/react';
import StyledHeader from '@layouts/Header';
import { BrowserRouter as Router } from 'react-router-dom';

export default {
  title: 'Header',
  component: StyledHeader,
} as ComponentMeta<typeof StyledHeader>;

const Template: ComponentStory<typeof StyledHeader> = () => (
  <Router>
    <StyledHeader />
  </Router>
);

export { Template };
