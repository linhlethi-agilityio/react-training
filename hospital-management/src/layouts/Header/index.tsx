import styled from 'styled-components';
import logo from '@image/logo-header.webp';
import { StyledSearchInput } from '@components/SearchInput';
import { StyledNavigation } from '@components/Navigation';
import { IHeaderProps } from './types';
import { NAVIGATION_ITEM } from '@constants/baseConfig';
import { ChangeEvent, memo, useState } from 'react';
import { IInputProps } from '@components/Input/types';
import { Link } from 'react-router-dom';
import StyledText from '@components/Text';
import { UserIcon } from '@components/Icons';
import { PAGE_ROUTES } from '@constants/routers';
import { StyledLink } from '@components/Link';

const Header = (props: IHeaderProps) => {
  const { className } = props;

  const [searchValue, setSearchValue] = useState('');
  const [locationValue, setLocationValue] = useState('');

  const onChangeLocation = (e: ChangeEvent<HTMLInputElement>) => {
    setLocationValue(e.target.value);
  };

  const onChangeSearch = (e: ChangeEvent<HTMLInputElement>) => {
    setSearchValue(e.target.value);
  };

  const locationInputConfig: IInputProps = {
    name: 'location',
    placeholder: 'Location',
    inputWidth: '80px',
    variant: 'primary',
  };

  const searchInputConfig: IInputProps = {
    name: 'search',
    placeholder: 'Search for Doctors, hospital & others....',
    inputWidth: '261px',
    variant: 'primary',
  };

  return (
    <header className={className}>
      <StyledLogo to="/">
        <StyledText as="h1">SwasthU</StyledText>
        <img src={logo} alt="logo" />
      </StyledLogo>
      <StyledSearchInput
        searchText={searchValue}
        location={locationValue}
        locationInputConfig={locationInputConfig}
        searchInputConfig={searchInputConfig}
        onChangeLocation={onChangeLocation}
        onChangeSearch={onChangeSearch}
      />
      <StyledNavigation items={NAVIGATION_ITEM} />
      <StyledLink to={`${PAGE_ROUTES.USERS}/1`}>
        <UserIcon />
      </StyledLink>
    </header>
  );
};

const StyledHeader = styled(Header)<IHeaderProps>`
  height: 100px;
  padding-left: 70px;
  padding-right: 36px;
  display: flex;
  align-items: center;
  justify-content: space-between;
  box-shadow: 0px 2px 20px rgba(0, 0, 0, 0.25);
`;

const StyledLogo = memo(
  styled(Link)`
    text-decoration: none;

    h1 {
      color: #ffffff;
    }

    img {
      position: absolute;
      z-index: 0;
      top: 13px;
    }
  `,
  () => true,
);

export default memo(StyledHeader);
