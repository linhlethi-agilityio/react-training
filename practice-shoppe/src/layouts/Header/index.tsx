import { memo, useCallback } from 'react';
import styled from 'styled-components';
import StyledText from '@components/Text';
import { StyledLink } from '@components/Link';
import { IHeaderProps } from './types';
import { ShoppingCartIcon } from '@components/Icons';
import { StyledButton } from '@components/Button';
import { useNavigate } from 'react-router-dom';
import { StyledLoadingSpinner } from '@components/LoadingSpinner';
import { StyledFlexBox } from '@components/FlexBox';
import { useAuth } from '@hooks/useAuth';

const logo = new URL('/images/logo.webp', import.meta.url).href;

const Header = (props: IHeaderProps) => {
  const { className } = props;

  const { data: user, isLoading, logout } = useAuth();
  const navigate = useNavigate();

  const renderButtonElement = useCallback(() => {
    if (user) {
      return (
        <StyledFlexBox gap={20}>
          <StyledLink to="/cart">
            <ShoppingCartIcon size={21} />
          </StyledLink>
          <StyledButton onClick={handleLogout} dis>
            logout
          </StyledButton>
          {isLoading && <StyledLoadingSpinner />}
        </StyledFlexBox>
      );
    } else
      return (
        <StyledLink to="/login">
          <StyledButton>login</StyledButton>
        </StyledLink>
      );
  }, [isLoading]);

  const handleLogout = useCallback(() => {
    logout();
    navigate('/login');
  }, []);

  return (
    <header className={className}>
      <StyledLogo to="/">
        <StyledText as="h1">Shoppe</StyledText>
        <img src={logo} alt="logo" />
      </StyledLogo>
      {renderButtonElement()}
    </header>
  );
};

const StyledLogo = memo(
  styled(StyledLink)`
    text-decoration: none;

    h1 {
      color: #ffffff;
    }

    img {
      position: absolute;
      z-index: 0;
      top: 60px;
    }
  `,
);

const StyledHeader = memo(styled(Header)`
  display: flex;
  justify-content: space-between;
  max-width: 1248px;
  padding-bottom: 10px;
  padding-right: 59px;
  padding-top: 66px;
  border-bottom: 1px solid #d8d8d8;
  margin: 0 auto;

  button {
    border: none;
    background: none;
  }
`);

export { StyledHeader };
