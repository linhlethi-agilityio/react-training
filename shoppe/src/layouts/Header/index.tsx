import { memo, useCallback, useContext } from 'react';
import styled from 'styled-components';
import StyledText from '@components/Text';
import { StyledLink } from '@components/Link';
import { IHeaderProps } from './types';
import { ShoppingCartIcon } from '@components/Icons';
import { UserContext } from '@contexts/UserContext';
import { StyledButton } from '@components/Button';
import { useNavigate } from 'react-router-dom';
import { StyledLoadingSpinner } from '@components/LoadingSpinner';
import { StyledFlexBox } from '@components/FlexBox';

const logo = new URL('/images/logo.webp', import.meta.url).href;

const Header = (props: IHeaderProps) => {
  const { className } = props;
  const { actions, state } = useContext(UserContext);

  const { loading } = state;
  const navigate = useNavigate();

  const renderButtonElement = useCallback(() => {
    const userLogin = localStorage.getItem('login');

    if (userLogin) {
      return (
        <StyledFlexBox gap={20}>
          <StyledLink to="/cart">
            <ShoppingCartIcon size={21} />
          </StyledLink>
          <StyledButton onClick={handleLogout} dis>
            logout
          </StyledButton>
          {loading && <StyledLoadingSpinner />}
        </StyledFlexBox>
      );
    } else
      return (
        <StyledLink to="/login">
          <StyledButton>login</StyledButton>
        </StyledLink>
      );
  }, [loading]);

  const handleLogout = useCallback(() => {
    actions.logout();
    localStorage.removeItem('login');
    navigate('/login');
  }, []);

  return (
    <header className={className}>
      <StyledLogo to="/">
        <StyledText as="h1">Shoppe</StyledText>
        <img src={logo} alt="logo" />
      </StyledLogo>
      {renderButtonElement()}
    </header>
  );
};

const StyledLogo = memo(
  styled(StyledLink)`
    text-decoration: none;

    h1 {
      color: #ffffff;
    }

    img {
      position: absolute;
      z-index: 0;
      top: 60px;
    }
  `,
);

const StyledHeader = memo(styled(Header)`
  display: flex;
  justify-content: space-between;
  max-width: 1248px;
  padding-bottom: 10px;
  padding-right: 59px;
  padding-top: 66px;
  border-bottom: 1px solid #d8d8d8;
  margin: 0 auto;

  button {
    border: none;
    background: none;
  }
`);

export { StyledHeader };
