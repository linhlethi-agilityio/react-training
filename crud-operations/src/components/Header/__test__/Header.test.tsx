import { render } from '@test-utils';
import '@testing-library/jest-dom';

// Components
import Header from '..';

const mockProps = {
  isClosedSideBar: false,
  onToggleSideBar: jest.fn(),
  onSearch: jest.fn(),
};

describe('Header component', () => {
  test('should match snapshot for Header', () => {
    const { container } = render(<Header {...mockProps} />);

    expect(container).toMatchSnapshot();
  });
});
