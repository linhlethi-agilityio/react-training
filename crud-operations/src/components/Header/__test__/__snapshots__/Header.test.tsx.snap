// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component should match snapshot for Header 1`] = `
<div>
  <div
    class="css-1yb29l5"
  >
    <button
      class="chakra-button css-4xx2wk"
      data-testid="toggle-sidebar-button"
      type="button"
    >
      <svg
        class="chakra-icon css-1njqkp8"
        focusable="false"
        height="18"
        viewBox="0 0 18 18"
        width="18"
      >
        <path
          d="M17.719 9A8.717 8.717 0 0 0 9 .281 8.717 8.717 0 0 0 .281 9 8.717 8.717 0 0 0 9 17.719 8.717 8.717 0 0 0 17.719 9ZM9 16.594c-4.173 0-7.594-3.379-7.594-7.594 0-4.173 3.379-7.594 7.594-7.594 4.173 0 7.594 3.379 7.594 7.594 0 4.173-3.379 7.594-7.594 7.594Zm1.125-4.219L6.75 9l3.375-3.375v6.75Zm1.125-6.75c0-.998-1.213-1.505-1.92-.795L5.955 8.205c-.439.44-.439 1.154 0 1.593l3.375 3.375c.707.707 1.92.207 1.92-.794V5.625Z"
          fill="#C4C4C4"
        />
      </svg>
    </button>
    <div
      class="css-1tmhq0v"
    >
      <div
        class="chakra-input__group css-10ip56e"
        data-group="true"
      >
        <input
          class="chakra-input css-2pjcr"
          placeholder="Search..."
          value=""
        />
        <div
          class="chakra-input__right-element css-dp9r5f"
          style="cursor: pointer;"
        >
          <svg
            height="14"
            viewBox="0 0 14 14"
            width="14"
          >
            <path
              d="M13.904 13.169 10.377 9.64a.325.325 0 0 0-.232-.095h-.282A5.687 5.687 0 0 0 5.688 0 5.687 5.687 0 0 0 0 5.687a5.687 5.687 0 0 0 9.546 4.179v.279c0 .087.035.17.095.232l3.528 3.527a.329.329 0 0 0 .465 0l.27-.27a.329.329 0 0 0 0-.465ZM5.688 10.5A4.81 4.81 0 0 1 .875 5.687 4.81 4.81 0 0 1 5.688.875 4.81 4.81 0 0 1 10.5 5.688 4.81 4.81 0 0 1 5.687 10.5Z"
              fill="#C4C4C4"
            />
          </svg>
        </div>
      </div>
      <button
        class="chakra-button css-4xx2wk"
        type="button"
      >
        <svg
          height="20"
          viewBox="0 0 17 20"
          width="17"
        >
          <path
            d="M8.5 18.75c.67 0 1.215-.562 1.215-1.251h1.214c0 1.38-1.09 2.501-2.429 2.501-1.338 0-2.428-1.122-2.428-2.501h1.214c0 .69.544 1.251 1.214 1.251ZM.555 13.079c1.061-1.04 1.872-2.129 1.872-5.817 0-3.11 2.406-5.645 5.466-5.951V.625C7.893.28 8.165 0 8.5 0c.336 0 .607.28.607.625v.686c3.06.307 5.466 2.842 5.466 5.95 0 3.689.812 4.778 1.873 5.817.53.52.697 1.306.427 2.002a1.811 1.811 0 0 1-1.694 1.17H1.822c-.753 0-1.418-.46-1.694-1.17a1.855 1.855 0 0 1 .427-2.002ZM1.822 15h13.357c.54 0 .81-.643.43-1.016-1.323-1.296-2.25-2.748-2.25-6.722C13.359 4.63 11.186 2.5 8.5 2.5c-2.685 0-4.858 2.13-4.858 4.762 0 3.959-.919 5.418-2.25 6.722-.382.374-.108 1.016.43 1.016Z"
            fill="#C4C4C4"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
